README.md
# wistroni40-bnft

# Install

```
npm i wistroni40-bnft --save
```

# Table of Contents

- [Quickstart](#quickstart)
- [Feature](#feature)
- [API](#api)
  - [BnftTemplate](#BnftTemplate)
  - [BenefitConfigModel](#BenefitConfigModel)
- [Resend API Usage](#resend-api-usage)

# Quickstart

> 繼承效益範本，實作`getBenefitParams`方法

benefit.ts

```typescript
import { BnftTemplate } from 'wistroni40-bnft';

/**
 * 效益使用範例
 */
class Benefit extends BnftTemplate {
  /**
   * 需要計算的廠別，若設為undefined則表示所有廠別都要計算
   */
  protected enabledPlant = ['plantA'];

  /**
   * @param config 效益設定檔
   */
  constructor(public config: BenefitConfigModel) {
    super(config);
  }

  /**
   * 取得效益參數
   *
   * @method public
   * @param condition 效益查詢條件
   * @return 回傳效益參數
   */
  public async getBenefitParams(
    condition: BenefitQueryModel,
  ): Promise<Bnft.Param[]> {
    const { site, plantCode } = condition;
    const idlCost = await this.findLatestLaborCost(
      site,
      plantCode,
      'idl',
    ).toPromise();

    const analysisParams: Bnft.Param = {
      name: 'time_analysis',
      value: 2,
      type: 'CONST',
    };

    const idlCostParams: Bnft.Param = {
      name: 'idl_pay_hr',
      value: idlCost ? idlCost.cost : 0,
      type: 'VAR',
    };

    return [analysisParams, idlCostParams];
  }
}
```

> 實例效益，使用`setSchedule`設定計算排成；或使用`execute`直接拋送效益

index.ts

```typescript
import { Benefit, BenefitConfigModel } from 'wistroni40-bnft';

// 效益設定
const configuration: BenefitConfigModel = {
  systemId: 'system_id',
  typeId: 'type_id',
  benefitType: 'direct',
  publishApi: 'http://publish-api-url/',
  benefitApi: 'http://benefit-api-url/',
  retry: 3,
};

const benefit = new Benefit(configuration);

// 設定排程定時拋送
benefit.setSchedule('0 0 0 * * ? *');

// 直接拋送
benefit.execute();

// 監聽拋送結果
benefit.sendCompleted.subscribe(res => {
  if (res.error) {
    // TODO
  } else {
    // TODO
  }
});
```

[Full Example](https://github.com/SteveLin100132/wistroni40-bnft/blob/master/examples/benefit.ts)

# Feature

- 提供效益計算實作範本
- 效益上拋具備重拋及保存機制
- 範本中埋入 API，可執行重新計算

# API

## **_BnftTemplate_**

**_Abstract Class_**，效益範本

### constructor

BnftTemplate 的建構值

| Parameter |        Type        | Required |   Default   | Description                                                       |
| :-------- | :----------------: | :------: | :---------: | :---------------------------------------------------------------- |
| config    | BenefitConfigModel | Required | `undefined` | 效益配置，配置說明可參考[BenefitConfigModel](#BenefitConfigModel) |

Example:

```typescript
const config = {
  systemId: 'mytest',
  typeId: 'cost_idl',
  benefitType: 'direct',
  publishApi: 'http://publish-api-url/',
  benefitApi: 'http://benefit-api-url/',
  retry: 3,
};
```

### Properties

#### config

|        Type        | Required |   Default   | Description |
| :----------------: | :------: | :---------: | :---------- |
| BenefitConfigModel | Required | `undefined` | 效益設定檔  |

### Methods

#### setSchedule

設定效益計算排程

| Parameter |     Type     | Required |   Default   | Description  |
| :-------- | :----------: | :------: | :---------: | :----------- |
| cron      |    string    | Required | `undefined` | 排程         |
| returns   | BnftTemplate | Required | `undefined` | 回傳物件本身 |

#### getBenefitParams

取得效益參數

| Parameter |            Type             | Required |   Default   | Description                                                           |
| :-------- | :-------------------------: | :------: | :---------: | :-------------------------------------------------------------------- |
| condition |      BenefitQueryModel      | Required | `undefined` | 效益查詢條件，配置說明可參考[BenefitConfigModel](#BenefitConfigModel) |
| returns   | Promise&lt;Bnft.Param[]&gt; | Required | `undefined` | 回傳效益參數                                                          |

#### processBenefitParams

處理效益參數

| Parameter |                              Type                               | Required |   Default   | Description        |
| :-------- | :-------------------------------------------------------------: | :------: | :---------: | :----------------- |
| response  |   Observable&lt;HttpResponse&lt;ActivedSystemModel[]&gt;&gt;    | Required | `undefined` | 查詢激活的效益系統 |
| timestamp |                              Date                               | Optional | `undefined` | 查詢開始時間       |
| returns   | Observable&lt;ProducePayloadModel&lt;Bnft.BenefitSaving&gt;&gt; | Required | `undefined` | 回傳處理後效益參數 |

#### execute

執行效益參數撈取

| Parameter |                              Type                               | Required |   Default   | Description          |
| :-------- | :-------------------------------------------------------------: | :------: | :---------: | :------------------- |
| timestamp |                              Date                               | Optional | `undefined` | 查詢開始時間         |
| returns   | Observable&lt;ProducePayloadModel&lt;Bnft.BenefitSaving&gt;&gt; | Required | `undefined` | 回傳效益參數上拋資料 |

#### send

將效益參數上拋

| Parameter |                     Type                      | Required |   Default   | Description      |
| :-------- | :-------------------------------------------: | :------: | :---------: | :--------------- |
| payload   | ProducePayloadModel&lt;Bnft.BenefitSaving&gt; | Required | `undefined` | 效益參數         |
| sendable  |                    boolean                    | Optional |    true     | 執行結果是否上拋 |

#### findLatestLaborCost

取得最新的 IDL 或 ID 人員工時

| Parameter |                       Type                       | Required |   Default   | Description                   |
| :-------- | :----------------------------------------------: | :------: | :---------: | :---------------------------- |
| site      |                      string                      | Required | `undefined` | Site                          |
| plantCode |                      string                      | Required | `undefined` | 廠別代碼                      |
| laborType |                'dl' &#124; 'idl'                 | Required | `undefined` | 人員類別(IDL or DL)           |
| returns   | Observable&lt;BenefitLatestLaborCostResponse&gt; | Required | `undefined` | 回傳最新的 IDL 或 ID 人員工時 |

#### setHttp

設定自定義的 HTTP

| Parameter |     Type     | Required |   Default   | Description      |
| :-------- | :----------: | :------: | :---------: | :--------------- |
| http      | HttpAdapter  | Required | `undefined` | 自定義 HTTP 請求 |
| returns   | BnftTemplate | Required | `undefined` | 回傳物件本身     |

## **_BenefitConfigModel_**

**_Interface_**，效益範本設定檔資料模型

### Properties

#### dev

|  Type   | Required | Default | Description |
| :-----: | :------: | :-----: | :---------- |
| boolean | Optional | `false` | 開發模式    |

#### systemId

|  Type  | Required |   Default   | Description |
| :----: | :------: | :---------: | :---------- |
| string | Required | `undefined` | 效益系統 ID |

#### typeId

|  Type  | Required |   Default   | Description |
| :----: | :------: | :---------: | :---------- |
| string | Required | `undefined` | 效益類型 ID |

#### benefitType

|    Type     | Required |   Default   | Description |
| :---------: | :------: | :---------: | :---------- |
| BenefitType | Required | `undefined` | 效益類型    |

#### publishApi

|  Type  | Required |   Default   | Description           |
| :----: | :------: | :---------: | :-------------------- |
| string | Required | `undefined` | 要拋送參數的 API 路徑 |

#### benefitApi

|  Type  | Required |   Default   | Description     |
| :----: | :------: | :---------: | :-------------- |
| string | Required | `undefined` | 效益的 API 路徑 |

#### retry

|  Type  | Required |   Default   | Description  |
| :----: | :------: | :---------: | :----------- |
| number | Optional | `undefined` | 重拋嘗試次數 |

# Resend API Usage

> 執行效益計算 Job，執行成功，可看見以下 LOG，並在瀏覽器輸
> 入`http://localhost:3000/`，可開啟 Swagger API Explorer

```log
[2021-02-17T10:30:24.562] [INFO] api - Listen api port 3000
```

## 重新上拋特定時間的所有效益

```
curl -X POST "http://localhost:3000/benefit/send?timestamp=1609430400000" -H  "accept: application/json"
```

| Paramter  |     Type      | Required |   Default   | Description    |
| :-------: | :-----------: | :------: | :---------: | :------------- |
| timestamp | query: number | Required | `undefined` | 重新上拋的時間 |

## 重新上拋特定效益

```
curl -X POST "http://localhost:3000/specific/benefit/send?systemId=mytest&typeId=cost_idl2&timestamp=1609430400000" -H  "accept: application/json"
```

| Paramter  |     Type      | Required |   Default   | Description    |
| :-------: | :-----------: | :------: | :---------: | :------------- |
| systemId  | query: string | Required | `undefined` | 系統 ID        |
|  typeId   | query: string | Required | `undefined` | 效益 ID        |
| timestamp | query: number | Required | `undefined` | 重新上拋的時間 |

## 重新上拋特定時間區間內的所有效益

```
curl -X POST "http://localhost:3000/benefit/send/timestamps" -H  "accept: application/json" -H  "Content-Type: application/json" -d "[1609430400000,1609516800000]"
```

| Paramter |   Type   | Required |   Default   | Description        |
| :------: | :------: | :------: | :---------: | :----------------- |
|   body   | number[] | Required | `undefined` | 重新上拋的時間區間 |

## 重新上拋特定時間區間內的特定效益

```
curl -X POST "http://localhost:3000/specific/benefit/send/timestamps?systemId=mytest&typeId=cost_idl2" -H  "accept: application/json" -H  "Content-Type: application/json" -d "[1609430400000]"
```

| Paramter |     Type      | Required |   Default   | Description        |
| :------: | :-----------: | :------: | :---------: | :----------------- |
| systemId | query: string | Required | `undefined` | 系統 ID            |
|  typeId  | query: string | Required | `undefined` | 效益 ID            |
|   body   |   number[]    | Required | `undefined` | 重新上拋的時間區間 |

## 重拋上拋失敗的效益

```
curl -X POST "http://localhost:3000/benefit/backup" -H  "accept: application/json"
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.7] - 2021-11-16

### Changed

- 效益上拋服務範本，`setSchedule` 及 `execute` 方法可訂閱上拋結果

## [1.1.6] - 2021-11-02

### Changed

- 效益上拋服務範本可設定設定檔客製欄位泛型

## [1.1.5] - 2021-06-09

### Changed

- 可以設定送出失敗嘗試重新拋送的時間間隔

## [1.1.4] - 2021-06-09

### Changed

- 監聽拋後結果是否成功，不管成功失敗，都會提供上拋數據完整資訊

## [1.1.3] - 2021-06-09

### Changed

- 開發模式送出效益資料時也能監聽拋送結果

## [1.1.2] - 2021-06-09

### Changed

- 效益範本設定添加開發模式，啟動開發模式不會送出效益僅打印日誌

## [1.1.0] - 2021-06-08

### Changed

- 效益設定資料提供客製設定欄位

## [1.0.9] - 2021-06-08

### Changed

- 上拋效益後，可以監聽拋後結果是否成功

## [1.0.8] - 2021-03-16

### Changed

- 匯出

## [1.0.7] - 2021-03-16

### Added

#### BnftTemplate

- 添加```enabledPlant```實作項目，來決定那些廠別是否要計算

## [1.0.6] - 2021-02-24

### Changed

#### BnftTemplate

- ```execute```方法添加log模式，執行時僅印出上拋參數，不進行數據拋送

## [1.0.5] - 2021-02-23

### Changed

- 更改Package跟目錄路徑

## [1.0.4] - 2021-02-22

### Changed

- 匯出

## [1.0.3] - 2021-02-22

### Added

- 效益背景API可終止

#### Resend Benefit API

## [1.0.2] - 2021-02-19

### Fixed

#### Resend Benefit API

- 效益背景API實現單例操作，避免多個效益服務產生多個後台API

#### Dependencies

- 將```swagger-jsdoc```, ```swagger-ui-express```加到```package.json```

## [1.0.1] - 2021-02-17

### Added

#### Resend Benefit API

- 提供重新上拋特定效益的API
- 提供重新上拋特定時間區間內的特定效益的API

### Changed

#### Resend Benefit API

- 支援多種效益在同一Job下執行API

## [1.0.0] - 2021-02-17

### Added

#### BnftTemplate

- 提供效益計算實作範本
- 設定效益計算排程
- 可查詢最新的IDL或ID人員工時
- 效益上拋具備重拋及保存機制

#### Resend Benefit API

- 提供重新上拋特定時間的所有效益的API
- 提供重新上拋特定時間區間內的所有效益的API
- 提供重拋上拋失敗的效益的API



src/api/server.ts
/**
 * 專案名稱： @wistroni40/bnft
 * 部門代號： ML8100
 * 檔案說明： API服務器
 * @CREATE Mon Feb 08 2021 下午2:56:08
 * @author Steve Y Lin
 * @contact Steve_Y_Lin@wistron.com #1342
 * -----------------------------------------------------------------------------
 * @NOTE
 */

import * as bodyParser from 'body-parser';
import * as express from 'express';
import * as fs from 'fs';
import * as http from 'http';
import { BnftTemplate } from 'src/bnft';
import * as swaggerUi from 'swagger-ui-express';
import { Log4js } from './../logger';
import { SWAGGER_DOC } from './swagger';

/**
 * API服務器
 */
export class Server {
  /**
   * 日誌
   */
  private readonly logger = new Log4js('api');
  /**
   * API Server
   */
  private readonly server = express();
  /**
   * API
   */
  private api?: http.Server;
  /**
   * 效益計算Job
   */
  private benefit = new Map<string, BnftTemplate>();
  /**
   * 單一實例
   */
  public static instance = new Server();

  private constructor() {
    // 啟動API Server
    this.start();
  }

  /**
   * 取得效益計算Job清單
   *
   * @method private
   * @return 回傳效益計算Job清單
   */
  private getBenefitList(): BnftTemplate[] {
    return Array.from(this.benefit).map(benefit => benefit[1]);
  }

  /**
   * 取得效益Key名稱
   *
   * @method private
   * @param req 請求
   * @return 回傳效益Key名稱
   */
  private getBenefitKey(req: express.Request): string {
    const systemId = req.query.systemId;
    const typeId = req.query.typeId;
    return `${systemId}.${typeId}`;
  }

  /**
   * 重新上拋所有效益
   *
   * @method private
   * @param req 請求
   * @param res 回應
   */
  private async sendBenefit(
    req: express.Request,
    res: express.Response,
  ): Promise<void> {
    const timestamp = new Date(parseInt(req.query.timestamp as string, 10));
    timestamp.setDate(timestamp.getDate() + 1);
    await Promise.all(
      this.getBenefitList().map(benefit =>
        benefit.execute(timestamp).toPromise(),
      ),
    );
    res.send(true);
  }

  /**
   * 重新上拋特定效益
   *
   * @method private
   * @param req 請求
   * @param res 回應
   */
  private async sendSpecificBenefit(
    req: express.Request,
    res: express.Response,
  ): Promise<void> {
    const key = this.getBenefitKey(req);
    const timestamp = new Date(parseInt(req.query.timestamp as string, 10));
    timestamp.setDate(timestamp.getDate() + 1);
    const benefit = this.benefit.get(key);
    if (benefit) {
      await benefit.execute(timestamp).toPromise();
      res.send(true);
    } else {
      res.send(null);
    }
  }

  /**
   * 重新上拋特定時間區間內的所有效益
   *
   * @method private
   * @param req 請求
   * @param res 會應
   */
  private async sendBenefitByTimeGroup(
    req: express.Request,
    res: express.Response,
  ): Promise<void> {
    if (Array.isArray(req.body) && req.body.length > 0) {
      await Promise.all(
        req.body.map(async date => {
          const timestamp = new Date(parseInt(date, 10));
          timestamp.setDate(timestamp.getDate() + 1);
          await Promise.all(
            this.getBenefitList().map(benefit =>
              benefit.execute(timestamp).toPromise(),
            ),
          );
        }),
      );
      res.send(true);
    } else {
      res.send('body format error');
    }
  }

  /**
   * 重新上拋特定時間區間內的特定效益
   *
   * @method private
   * @param req 請求
   * @param res 會應
   */
  public async sendSpecificBenefitByTimeGroup(
    req: express.Request,
    res: express.Response,
  ): Promise<void> {
    if (Array.isArray(req.body) && req.body.length > 0) {
      const key = this.getBenefitKey(req);
      await Promise.all(
        req.body.map(async date => {
          const timestamp = new Date(parseInt(date, 10));
          timestamp.setDate(timestamp.getDate() + 1);
          const benefit = await this.benefit.get(key);
          if (benefit) {
            await benefit.execute(timestamp).toPromise();
          }
        }),
      );
      res.send(true);
    } else {
      res.send('body format error');
    }
  }

  /**
   * 重拋上拋失敗的效益
   *
   * @method private
   * @param req 請求
   * @param res 會應
   */
  private async sendBackupBenefit(
    req: express.Request,
    res: express.Response,
  ): Promise<void> {
    const files = fs.readdirSync('backup');
    await Promise.all(
      files.map(file => {
        const benefit = fs.readFileSync(`backup/${file}`, { encoding: 'utf8' });
        try {
          if (this.getBenefitList().length > 0) {
            this.getBenefitList()[0].send(JSON.parse(benefit));
            fs.unlinkSync(`backup/${file}`);
          }
        } catch (error) {
          this.logger.warn('Send backup benefit failed');
          this.logger.warn(benefit);
        }
        return JSON.parse(benefit);
      }),
    );
    res.send(true);
  }

  /**
   * 啟動API Server
   *
   * @method public
   * @return 回傳物件本身
   */
  public start(): Server {
    const port = 3000;
    this.server.use(bodyParser.json());
    this.server.post('/benefit/send', this.sendBenefit.bind(this));
    this.server.post(
      '/specific/benefit/send',
      this.sendSpecificBenefit.bind(this),
    );
    this.server.post(
      '/benefit/send/timestamps',
      this.sendBenefitByTimeGroup.bind(this),
    );
    this.server.post(
      '/specific/benefit/send/timestamps',
      this.sendSpecificBenefitByTimeGroup.bind(this),
    );
    this.server.post('/benefit/backup', this.sendBackupBenefit.bind(this));
    this.server.use('/', swaggerUi.serve, swaggerUi.setup(SWAGGER_DOC));
    this.logger.info(`Listen api port ${port}`);
    this.api = this.server.listen(port);
    return this;
  }

  /**
   * 終止API Server
   *
   * @method public
   * @return 回傳物件本身
   */
  public stop(): Server {
    if (this.api) {
      this.api.close();
      this.logger.warn('api close');
    }
    return this;
  }

  /**
   * 註冊效益計算Job
   *
   * @method public
   * @param benefit 效益計算Job
   */
  public register(benefit: BnftTemplate): void {
    const key = `${benefit.config.systemId}.${benefit.config.typeId}`;
    this.benefit.set(key, benefit);
  }
}
